File: java/src/algorithms/sorting/merge_sort/MergeSort.java
package algorithms.sorting.merge_sort;

import algorithms.common.Cargo;
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;

public class MergeSort {
    /**
     * Sorts an ArrayList of Cargo objects using Merge Sort
     * Time Complexity: O(n log n)
     * 
     * @param items ArrayList to sort
     * @return time taken in nanoseconds
     */
    public static long sort(ArrayList<Cargo> items) {
        long start = System.nanoTime();

        ArrayList<Cargo> sorted = mergeSort(items);
        // Copy sorted items back to original list
        for (int i = 0; i < items.size(); i++) {
            items.set(i, sorted.get(i));
        }

        return System.nanoTime() - start;
    }

    private static ArrayList<Cargo> mergeSort(ArrayList<Cargo> items) {
        if (items.size() <= 1) {
            return items;
        }

        int mid = items.size() / 2;
        List<Cargo> left = new LinkedList<>(items.subList(0, mid));
        List<Cargo> right = new LinkedList<>(items.subList(mid, items.size()));

        return merge(mergeSort(new ArrayList<>(left)),
                mergeSort(new ArrayList<>(right)));
    }

    private static ArrayList<Cargo> merge(ArrayList<Cargo> left, ArrayList<Cargo> right) {
        ArrayList<Cargo> merged = new ArrayList<>();
        int leftIndex = 0, rightIndex = 0;

        while (leftIndex < left.size() && rightIndex < right.size()) {
            if (left.get(leftIndex).compareTo(right.get(rightIndex)) < 0) {
                merged.add(left.get(leftIndex++));
            } else {
                merged.add(right.get(rightIndex++));
            }
        }

        while (leftIndex < left.size()) {
            merged.add(left.get(leftIndex++));
        }

        while (rightIndex < right.size()) {
            merged.add(right.get(rightIndex++));
        }

        return merged;
    }

    /**
     * Verify if the ArrayList is properly sorted
     * 
     * @param items ArrayList to check
     * @return true if sorted, false otherwise
     */
    public static boolean isSorted(ArrayList<Cargo> items) {
        if (items == null || items.isEmpty() || items.size() == 1) {
            return true;
        }

        for (int i = 1; i < items.size(); i++) {
            if (items.get(i - 1).compareTo(items.get(i)) > 0) {
                return false;
            }
        }
        return true;
    }
}

File: python/algorithms/sorting/merge_sort/merge_sort.py
"""
Merge Sort Algorithm Implementation
Author: Bjorn Melin
Date: 1/28/2025

Merge Sort is a divide-and-conquer algorithm that divides the input array into 
two halves, recursively sorts them, and then merges the sorted halves to produce 
the final sorted array.

Time Complexity:
- Best Case: O(n log n)
- Average Case: O(n log n)
- Worst Case: O(n log n)

Space Complexity: O(n)
"""

from typing import List


def merge_sort(arr: List[int]) -> List[int]:
    """
    Sorts an array of integers using the Merge Sort algorithm.

    Args:
        arr (List[int]): The list of integers to be sorted.

    Returns:
        List[int]: The sorted list of integers.
    """
    if len(arr) <= 1:
        return arr

    mid = len(arr) // 2
    left_half = merge_sort(arr[:mid])
    right_half = merge_sort(arr[mid:])

    return merge(left_half, right_half)


def merge(left: List[int], right: List[int]) -> List[int]:
    """
    Merges two sorted lists into one sorted list.

    Args:
        left (List[int]): The left sorted list.
        right (List[int]): The right sorted list.

    Returns:
        List[int]: The merged sorted list.
    """
    sorted_list = []
    left_index, right_index = 0, 0

    while left_index < len(left) and right_index < len(right):
        if left[left_index] < right[right_index]:
            sorted_list.append(left[left_index])
            left_index += 1
        else:
            sorted_list.append(right[right_index])
            right_index += 1

    sorted_list.extend(left[left_index:])
    sorted_list.extend(right[right_index:])

    return sorted_list
